Deconstructive Activities: Rework contract data enrichment

Shadowfiend opened the pull request:
Few things here:
- Contract data enrichment is now handled via a stub for a future enrichment
  service. The stub is just a function.
- The activities slice now tracks transactions with null block heights, meaning
  pending transactions.
- Contract info is now added to a separate field in the `ActivityItem` type.
  The activity list items have a harness for treating different resolved
  contract info differently, heavily based on the work in #537 but meant to
  illustrate an idea of a future direction. It's still a bit dirty and could
  use a little more abstraction in the future.
- The accounts slice no longer tracks any transaction activity. The existing
  data was unused anyway.
- The activities reducer is now typed correctly and the `infoRows` it adds are
  now handled correctly in the UI. They were already rendering roughly
  correctly, but it was an accident rather than intentional proper use of the
  data.

mhluongo commented on background/redux-slices/activities.ts:
```diff
@@ -1,23 +1,28 @@
   sortComparer: (a, b) => {
-    if (a.blockHeight === b.blockHeight) {
+    // null means pending, pending is always sorted above everything.
+    if (a.blockHeight === null && b.blockHeight === 0) {
```
... what if they're both `null`? Is that `0` a mistake?

Shadowfiend replied:
... Yep 😅

mhluongo commented on background/redux-slices/utils/activity-utils.ts:
```diff
@@ -110,39 +162,76 @@ export const keysMap: UIAdaptationMap<ActivityItem> = {
+      contractLogoURL,
+      assetAmount: enrichAssetAmountWithDecimalValues(
+        {
+          asset: matchingFungibleAsset,
```
Note that if this tx is a `transfer()` that internally moves ERC-20 tokens around, this `matchingFungibleAsset` might not be the same as the contract that has has `transfer()` called on it. Might want to confirm `matchingFungibleAsset`'s contract address is the same as the `to` of the tx, otherwise return a general contract interaction

Shadowfiend replied:
You're saying if the resolved target is the same as the sender? The matching asset is resolved against the transaction to address.

mhluongo replied:
Oh I'd assumed we got the asset from Alchemy metadata. If it's already from the transaction `to` this isn't an issue ofc

mhluongo commented on background/redux-slices/utils/activity-utils.ts:
```diff
@@ -110,39 +162,76 @@ export const keysMap: UIAdaptationMap<ActivityItem> = {
+    // a simple ETH send to a contract address can still effectively be a
+    // contract interaction (because it calls the fallback function on the
+    // contract), but for now we deliberately ignore that scenario when
+    // categorizing activities.
```
This is a beautiful comment. I'd love to see it for any similarly intentional stuff about, eg like including the `contractLogoURL` with a general `contract-interaction`

mhluongo commented on background/redux-slices/utils/activity-utils.ts:
```diff
@@ -110,39 +162,76 @@ export const keysMap: UIAdaptationMap<ActivityItem> = {
+
+  return {
+    ...transaction,
+    contractInfo: resolveContractInfo(
```
I can't believe this isn't `async`, we have so much more to do here heh

Shadowfiend replied:
Imo this shouldn't need to be async even when we're looking up more stuff--it should just emit an updated thing once we resolve whatever else we want to resolve.

mhluongo replied:
That's two steps ahead 😁

mhluongo reviewed (commented):
Left a couple thoughts. Tested locally, and clearly this thing hasn't been styled yet.... it's a draft, after all! But I do see some of my erroneous tx labelling has been replaced by "contract interaction" which is nice.

mhluongo reviewed (approved):
I'm happy with this but looking for another :+1: before merge. @henryboldi?

henryboldi reviewed (changes requested):
This is amazing! 🧙‍♂️ I've got a couple questions and things I'd like to flag.

1) I believe a contract interaction shouldn't be listed as a sent asset in the activity detail panel

<img width="340" alt="contract interaction not displayed" src="https://user-images.githubusercontent.com/1918798/145458634-460ffe95-f5bc-4cf0-81fa-03ceb560bb41.png">

2) To be clear, we're not marking swaps in this PR?

3) Does this change retain the token data for displaying non-ETH transactions? It wasn't exhaustive but we used to be able to display (in #537),
<img width="377" alt="Zero" src="https://user-images.githubusercontent.com/1918798/145459066-6489103f-f936-41b4-a4e0-7b98462c078d.png">
Is that not worth keeping?

Shadowfiend commented on background/main.ts:
```diff
@@ -289,24 +287,19 @@ export default class Main extends BaseService<never> {
+      const enrichedTransaction = enrichTransactionWithContractInfo(
+        this.store.getState().assets,
+        transaction,
+        2 /* TODO desiredDecimals should be configurable */
```
It's all string amounts for formatting.

Shadowfiend commented on background/services/enrichment/index.ts:
```diff
@@ -0,0 +1,129 @@
+      assetAmount: enrichAssetAmountWithDecimalValues(
+        {
+          asset: matchingFungibleAsset,
+          amount: BigInt(`0x${contractInput.slice(10, 10 + 64)}`),
```
Yep, see the TODO above.

Shadowfiend commented:
> To be clear, we're not marking swaps in this PR?

Correct. We have the infra for it, but not the identification or full rendering for now.

> I believe a contract interaction shouldn't be listed as a sent asset in the activity detail panel

Whoops, failed to pull in your header adjustments. Fixed.

> Does this change retain the token data for displaying non-ETH transactions?

Only when we are reasonably certain that it's an ERC20 token transfer (by checking the function selector and input length for something we've identified as a smart contract fungible asset). Otherwise we don't actually know the data we're reading is correct, because we have no idea what the function called in the transaction is or what its input ABI looks like—we're just offsetting into input data and hoping the function being called takes a single uint256 that is a token amount, and that this means it's a token transfer. That is the salient piece of this identifier: https://github.com/tallycash/extension/pull/538/commits/6b69209b6d51ec032537ed0b9670669766e94809#diff-e21d963357c1054efb4597ad83510c25be9ca92fd5280c0b53d0c2ab945466fbR190-R192 .

As an example, as far as I can tell, the ZERO transaction in that list isn't correct. That transaction resulted in 100 ZERO being minted, whereas 0.1 is the ETH sent with the transaction. If there had been a 0 ETH value in the transaction, we would have tried to pull the value from the first parameter to the function call—but the function called in that transaction has as its first parameter an address, not an amount, so we would have pulled the amount 1.410102024832666e+48 or so (after dividing out the 18 decimals). The amount that particular function takes is the third parameter—and it's actually (for reasons that are unclear to me, since this info is available without a function parameter) the ETH amount sent with the transaction.

The potential to display some transactions incorrectly is a greater problem than the utility of displaying some other ones more comprehensively. This PR's goal was to set us up for a good way to comprehensively categorize transactions based on deeper understanding of their input structure.

**With all of that said**… The ERC20 recognizer was wrong as-was, and I absolutely would not have noticed if you hadn't asked 😅 It was checking for too short an input length (`transfer` transaction parameters are 128 characters long), and it was trying to pull the value from the first rather than second parameter (the first 64 characters are the recipient address). I've pushed a fix 🙌 

While fixing this, I realized that the first transaction inserted into an account always gets lost due to something weird in how `upsertOne` works with the result of `getInitialState`. I also realized that we were hiding 0-ETH activities, which will lose all asset transfers in the list. I fixed both of those.

Last but not least, I've added recipient information to the asset transfer so that the To in the activity list is correctly the target of the transfer.

You can test this one by adding account 0x2a8405140dc94965478c51c522d84615ff96dc92, whose latest transaction is a UNI transfer.

mhluongo commented on background/services/enrichment/index.ts:
```diff
@@ -0,0 +1,131 @@
+  // FIXME Move to ERC20 parsing using ethers.
+  if (
+    typeof matchingFungibleAsset !== "undefined" &&
+    contractInput.length === 138 &&
```
:ok_hand:

mhluongo reviewed (approved):
Transactions look good on my local wallet. Can't wait to write more tx recognizers. Let's get this merged!
